### PR TITLE
fs: move, movedir and rename implementation

### PR DESCRIPTION
**Pull request by lnielsen**
* Adds support for move, movedir and rename methods. (addresses #9)

* Fixes bug in XRootDFS._stat_info which did not access the right
  attribute.

* Fixes issue with removedir not being able to remove a non-empty
  directory.

Signed-off-by: Lars Holm Nielsen <lars.holm.nielsen@cern.ch>

### CODE REVIEW

File: xrootdfs/fs.py [hunk 1]
```diff
@@ -372,3 +395,68 @@ def wildcard(fn):
...
+        """Move a directory from one location to another.
+
+        :param src: source directory path
+        :type src: string
+        :param dst: destination directory path
+        :type dst: string
+        :param overwrite: if True then any existing files in the destination
+            directory will be overwritten
+        :type overwrite: bool
+        :param ignore_errors: if True then this method will ignore FSError
```
Comment (lnielsen) [inline xrootdfs/fs.py]: Fix documentation strings

File: xrootdfs/fs.py [hunk 2]
```diff
@@ -258,9 +259,26 @@ def removedir(self, path, recursive=False, force=False):
...
         """
         status, res = self.client.rmdir(self._p(path))
-        if not status.ok:
-            raise FSError(details=status)
-        return True
+
+        if status.ok:
+            return True
+
+        if force and status.errno == 3005:
```
Comment (lnielsen) [inline xrootdfs/fs.py]: Tests for removedir will be added in a separate pull request.

**Comment by lnielsen** [comment]
@otron can you take a look at this PR.

File: tests/test_fs.py [hunk 3]
```diff
@@ -178,13 +179,203 @@ def test_open(tmppath):
...
     xrd_rooturl = mkurl(tmppath)
 
-    full_file_path = '/' + join(tmppath, file_name)
-
     # Open file w/ xrootd
     xrdfs = XRootDFS(xrd_rooturl)
     xfile = xrdfs.open(file_name, mode='r')
     assert xfile
     assert type(xfile) == XRootDFile
+    xfile.read() == contents
```
Comment (otron) [inline tests/test_fs.py]: Missing `assert`?

File: xrootdfs/fs.py [hunk 4]
```diff
@@ -145,7 +146,7 @@ def _stat_flags(self, path):
         status, stat = self.client.stat(self._p(path))
 
         if not status.ok:
```
Comment (otron) [inline xrootdfs/fs.py]: To match the syntax in `xrdfile.py` it should be `if not status.ok or status.error:` (ref #7)
I don't know if it's necessary to check both, as I've not found any mention of their relationship in the pyxrootd docs.

File: xrootdfs/fs.py [hunk 5]
```diff
@@ -258,9 +259,26 @@ def removedir(self, path, recursive=False, force=False):
         :raises `fs.errors.ResourceNotFoundError`: if the path does not exist
         """
         status, res = self.client.rmdir(self._p(path))
-        if not status.ok:
-            raise FSError(details=status)
-        return True
+
+        if status.ok:
```
Comment (otron) [inline xrootdfs/fs.py]: ref. #7

File: xrootdfs/fs.py [hunk 6]
```diff
@@ -258,9 +259,26 @@ def removedir(self, path, recursive=False, force=False):
...
+            # it ourselves.
+            for d, filenames in self.walk(path, search="depth"):
+                for filename in filenames:
+                    path = pathjoin(d, filename)
+                    status, res = self.client.rm(
+                        path, timeout=self.timeout)
+                    if not status.ok:
+                        self._raise_status(path, status)
+                status, res = self.client.rmdir(d, timeout=self.timeout)
+                if not status.ok:
```
Comment (otron) [inline xrootdfs/fs.py]: ref #7

File: xrootdfs/fs.py [hunk 7]
```diff
@@ -258,9 +259,26 @@ def removedir(self, path, recursive=False, force=False):
...
+
+        if force and status.errno == 3005:
+            # pyxrootd does not support recursive removal so do we have to do
+            # it ourselves.
+            for d, filenames in self.walk(path, search="depth"):
+                for filename in filenames:
+                    path = pathjoin(d, filename)
+                    status, res = self.client.rm(
+                        path, timeout=self.timeout)
+                    if not status.ok:
```
Comment (otron) [inline xrootdfs/fs.py]: ref #7

File: xrootdfs/fs.py [hunk 8]
```diff
@@ -372,3 +395,68 @@ def wildcard(fn):
...
+        """Internal move function."""
+        if overwrite and self.exists(dst):
+            if self.isfile(dst):
+                self.remove(dst)
+            elif self.isdir(dst):
+                self.removedir(dst, force=True)
+
+        status, dummy = self.client.mv(src, dst, timeout=self.timeout)
+
+        if not status.ok:
```
Comment (otron) [inline xrootdfs/fs.py]: ref #7

**Comment by otron** [comment]
Besides that it looks good!

File: xrootdfs/fs.py (hunk 4 above)
Comment (lnielsen) [inline xrootdfs/fs.py]: I think we should just check against ``status.ok``. I assume that ``status.error`` should like be considered together with ``status.fatal`` to determine how bad the error is, but we should ask Elvin instead.

Comment (otron) [inline xrootdfs/fs.py]: :+1: for asking Elvin :)

Comment (lnielsen) [inline xrootdfs/fs.py]: Got a reply from Elvin - we should just check ``status.ok``
